### PR TITLE
add NDJSON support

### PR DESCRIPTION
**Pull request by lei0zhou**
this pr is an implementation for #18

### CODE REVIEW

File: lib/epiviewpoint/data_file.ex [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    read(string, header_transformer, headers, &parse_csv/1)
+  end
+
+  def read(string, :ndjson, header_transformer, headers) when is_binary(string) do
+    read(string, header_transformer, headers, &parse_ndjson/1)
+  end
+
+  def read(input, header_transformer, [required: required_headers, optional: optional_headers], parser) do
+    with {:ok, df} <- parser.(input),
+         provided_headers <- df |> DataFrame.names() |> Enum.map(&String.trim/1) |> header_transformer.(),
```
Comment (jc00ke) [inline lib/epiviewpoint/data_file.ex:26]: Extract to a private function. I prefer for `with` to be clean and readable.

File: test/epiviewpoint_web/controllers/import_controller_test.exs [hunk 2]
```diff
@@ -87,14 +87,68 @@ defmodule EpiViewpointWeb.ImportControllerTest do
       assert conn |> redirected_to() == "/import/start"
       assert "Invalid mm-dd-yyyy format: 06/02/bb" = Session.get_import_error_message(conn)
     end
+
+    test "accepts ndjson file upload", %{conn: conn} do
+      temp_file_path =
+        """
+        {"search_firstname_2":"Alice","search_lastname_1":"Testuser","dateofbirth_8":"01/01/1970","datecollected_36":"06/02/2020","resultdate_42":"06/01/2020","datereportedtolhd_44":"06/03/2020","result_39":"positive","glorp":"393","person_tid":"alice"}
```
Comment (jc00ke) [inline test/epiviewpoint_web/controllers/import_controller_test.exs:94]: Should there be this blank line?

File: test/epiviewpoint_web/controllers/import_controller_test.exs [hunk 3]
```diff
@@ -87,14 +87,68 @@ defmodule EpiViewpointWeb.ImportControllerTest do
...
+        {"search_firstname_2":"Billy","search_lastname_1":"Testuser","dateofbirth_8":"03/01/1990","datecollected_36":"06/05/2020","resultdate_42":"06/06/2020","datereportedtolhd_44":"06/07/2020","result_39":"negative","glorp":"sn3","person_tid":"billy"}
+        """
+        |> Tempfile.write_ndjson!()
```
Comment (jc00ke) [inline test/epiviewpoint_web/controllers/import_controller_test.exs:97]: Consider using [ExUnit's tmp dir](https://hexdocs.pm/ex_unit/ExUnit.Case.html#module-tmp-dir) and then you don't need the `File.rm/1`.

File: test/epiviewpoint_web/controllers/import_controller_test.exs [hunk 4]
```diff
@@ -87,14 +87,68 @@ defmodule EpiViewpointWeb.ImportControllerTest do
...
+               total_lab_result_count: 2,
+               total_person_count: 2
+             } = Session.get_last_file_import_info(conn)
+    end
+
+    test "when a required field is missing in ndjson", %{conn: conn} do
+      # remove the dateofbirth_8 field
+      temp_file_path =
+        """
+        {"search_firstname_2":"Alice","search_lastname_1":"Testuser","datecollected_36":"06/02/2020","resultdate_42":"06/01/2020","datereportedtolhd_44":"06/03/2020","result_39":"positive","glorp":"393","person_tid":"alice"}
```
Comment (jc00ke) [inline test/epiviewpoint_web/controllers/import_controller_test.exs:117]: Newline?

File: test/epiviewpoint_web/controllers/import_controller_test.exs [hunk 5]
```diff
@@ -87,14 +87,68 @@ defmodule EpiViewpointWeb.ImportControllerTest do
...
+        """
+        |> Tempfile.write_ndjson!()
```
Comment (jc00ke) [inline test/epiviewpoint_web/controllers/import_controller_test.exs:119]: `@tmp_dir`?

**Review by jc00ke — CHANGES_REQUESTED** [review]

File: test/epiviewpoint_web/controllers/import_controller_test.exs (hunk 2 above)
Comment (lei0zhou) [reply test/epiviewpoint_web/controllers/import_controller_test.exs:94]: not sure if I understand correctly, the `"""` will not create newline at the beginning, as opposed to `"`

image

File: .gitignore
[the remark this answers is not in the document]
Comment (lei0zhou) [reply .gitignore:1]: removed. will make sure the scope next time

File: elixir_buildpack.config
[the remark this answers is not in the document]
Comment (lei0zhou) [reply elixir_buildpack.config:1]: removed.

File: test/epiviewpoint_web/controllers/import_controller_test.exs (hunk 3 above)
Comment (lei0zhou) [reply test/epiviewpoint_web/controllers/import_controller_test.exs]: updated.

File: lib/epiviewpoint/data_file.ex (hunk 1 above)
Comment (lei0zhou) [reply lib/epiviewpoint/data_file.ex:26]: sorry I missed this, will update soon

Comment (lei0zhou) [reply lib/epiviewpoint/data_file.ex]: updated.

File: test/epiviewpoint_web/controllers/import_controller_test.exs [hunk 6]
```diff
@@ -11,7 +11,9 @@ defmodule EpiViewpointWeb.ImportControllerTest do
   @admin Test.Fixtures.admin()
 
   describe "create" do
-    test "prevents non-admins from uploading", %{conn: conn, user: user} do
+    @describetag :tmp_dir
```
Comment (jc00ke) [inline test/epiviewpoint_web/controllers/import_controller_test.exs:14]: I've not seen [@describetag](https://hexdocs.pm/ex_unit/ExUnit.Case.html#module-module-and-describe-tags) before, very cool!

File: test/epiviewpoint_web/controllers/import_controller_test.exs (hunk 2 above)
Comment (jc00ke) [reply test/epiviewpoint_web/controllers/import_controller_test.exs:88]: image

Line 88 looks like a blank line but maybe it's just how it's rendered?

**Review by jc00ke — APPROVED** [review]

**Pull request merged**